src/mscorlib/shared/System/SpanHelpers.Char.cs

```diff
@@ -84,79 +84,87 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
-            {
-                nLength -= 4;
-
-                if (uValue == Unsafe.Add(ref searchSpace, index))
-                    goto Found;
-                if (uValue == Unsafe.Add(ref searchSpace, index + 1))
-                    goto Found1;
-                if (uValue == Unsafe.Add(ref searchSpace, index + 2))
-                    goto Found2;
-                if (uValue == Unsafe.Add(ref searchSpace, index + 3))
-                    goto Found3;
-
-                index += 4;
-            }
+                while (length >= 4)
+                {
+                    length -= 4;
+
+                    if (*pCh == value)
+                        goto Found;
+                    if (*(pCh + 1) == value)
+                        goto Found1;
+                    if (*(pCh + 2) == value)
+                        goto Found2;
+                    if (*(pCh + 3) == value)
+                        goto Found3;
+
+                    pCh += 4;
+                }
 
-            while ((byte*)nLength > (byte*)0)
-            {
-                nLength -= 1;
+                while (length > 0)
+                {
+                    length -= 1;
 
-                if (uValue == Unsafe.Add(ref searchSpace, index))
-                    goto Found;
+                    if (*pCh == value)
+                        goto Found;
 
-                index += 1;
-            }
+                    pCh += 1;
+                }
 #if !netstandard11
-            if (Vector.IsHardwareAccelerated && ((int)(byte*)index < length))
-            {
-                nLength = (IntPtr)((length - (int)(byte*)index) & ~(Vector<ushort>.Count - 1));
+                // We get past SequentialScan only if IsHardwareAccelerated is true. However, we still have the redundant check to allow
+                // the JIT to see that the code is unreachable and eliminate it when the platform does not have hardware accelerated.
+                if (Vector.IsHardwareAccelerated && pCh < pEndCh)
+                {
+                    length = (int)((pEndCh - pCh) & ~(Vector<ushort>.Count - 1));
 
-                // Get comparison Vector
-                Vector<ushort> vComparison = new Vector<ushort>(value);
+                    // Get comparison Vector
+                    Vector<ushort> vComparison = new Vector<ushort>(value);
 
-                while ((byte*)nLength > (byte*)index)
-                {
-                    var vMatches = Vector.Equals(vComparison, Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref searchSpace, index))));
-                    if (Vector<ushort>.Zero.Equals(vMatches))
+                    while (length > 0)
                     {
-                        index += Vector<ushort>.Count;
-                        continue;
+                        // Using Unsafe.Read instead of ReadUnaligned since the search space is pinned and pCh is always word aligned
+                        Debug.Assert(((int)pCh % IntPtr.Size) == 0);
```

Comment:
This assert should be use `sizeof(Vector<byte>)` instead of `IntPtr.Size`.

Reply:
`sizeof(Vector<byte>)` -> `Vector<byte>.Count`?

Reply:
It should be `sizeof(Vector<ushort>)` since we are dealing with `Vector<ushort>` here.

`Vector<byte>.Count` happens to have the same value as `sizeof(Vector<ushort>)` today, but it is less correct.

Reply:
I can't use `sizeof` here:
` error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('Vector<ushort>')`

Reply:
So use `Unsafe.SizeOf<Vector<ushort>>()`